### ISSUE TITLE
(change) Allow empty evaluation context for some flags.

### ISSUE THREAD

**Issue by thomaspoignant** [thomaspoignant/go-feature-flag#2533]
### Motivation

As of today all flag evaluation are needing at least a `targetingKey`.
The reason behind that is because the `targetingKey` is used to do bucketing during the flag evaluation _(specially for percentage based rules)_

But not all flags requires bucketing, so in some cases it should be ok, not to have a bucketing key (`targetingKey` or any other `bucketingKey`).

### Example
```yaml
my-new-feature:
  variations:
    enabled: true
    disabled: false
  defaultRule:
    variation: disabled
```

A flag like this one does not require a bucketing key, so we should accept an empty evaluation context for such flag. 

### Requirements

- We should allow empty evaluation context for flag that does not require it.
- In case of the flag requires a bucketing key and has no `targetingKey` we should not error if the flag is configured to use a different `bucketingKey`.
- In case of the flag use a percentage based rollout and that we don't have any key to do the variant affection we should error with `TARGETING_KEY_MISSING`.
- Today all OF providers are checking the presence of the `targetingKey`, we should remove this mandatory thing and let the evaluation goes to the relay-proxy before returning any errors.

_This change is quite impactful and touch a lot of core components of GO Feature Flag, if you want to work on it please reach out before jumping on the implementation._

**Comment by bruhhgnik** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/go-feature-flag).

### My background and how it can be leveraged
Hi im nik, a beginner blockchain and linux dev

### How I plan on tackling this issue
I would love to:

Modify flag evaluation to support optional targeting keys
Update providers to remove mandatory targetingKey check
Add validation for percentage-based rollouts
Implement new error handling for missing bucketing keys
Handle empty evaluation contexts for simple flags"

ETA:
5-6 days 

Since this touches core functionality and requires careful testing, we need to be conservative with the timeline. Each change needs thorough validation to avoid breaking existing implementations

**Comment by martinvibes** [comment]
hello @thomaspoignant can i work on this

**Comment by seunghunlee01** [comment]
Hi @thomaspoignant I met the same issue.
could you please assign this issue to @martinvibes or the other member?

**Comment by thomaspoignant** [comment]
@seunghunlee01 do you want to work on this?

**Comment by seunghunlee01** [comment]
> @seunghunlee01 do you want to work on this?

Hi @thomaspoignant , I don't know well golang and this open source.
but If you're ok, then I want to work it.

**Comment by thomaspoignant** [comment]
This is the not the easiest issue, but if you feel that you can do it your welcome.

**Comment by seunghunlee01** [comment]
Thank you, I'll try this task.
If It's too hard for me, then I'll tell you again.

**Comment by thomaspoignant** [comment]
A way to address this is to check if the flag need a bucketing key.
Bucketing keys are useful for percentage and progressive rollout, so a function like this one can tell if we need a bucketing key or not:

```go
func (f *InternalFlag) flagNeedBucketingKey() bool {
	if f.DefaultRule.Percentages != nil {
		return true
	}
	if f.DefaultRule.ProgressiveRollout != nil {
		return true
	}

	for _, rule := range f.GetRules() {
		if rule.Disable != nil && *rule.Disable {
			continue
		}
		if rule.Percentages != nil && len(rule.GetPercentages()) > 0 {
			return true
		}
		if rule.ProgressiveRollout != nil {
			return true
		}
	}
	
	return false
}
```

**Comment by mimisavage** [comment]
May I try my hand at this?

**Comment by thomaspoignant** [comment]
@mimisavage yes you can follow the contributing guidelines and add it to yourself.

**Comment by Volodymyr-Maiboroda-Imprivata** [comment]
@thomaspoignant is there any plan to implement it soon?

**Comment by thomaspoignant** [comment]
@Volodymyr-Maiboroda-Imprivata I haven’t worked on it yet.
But this is part of what we want to provide.

**Comment by Volodymyr-Maiboroda-Imprivata** [comment]
@thomaspoignant, we are currently encountering an issue related to feature flag evaluation, where we do not have a dedicated target key available. In our setup, the target key property is shared across multiple users, which appears to be causing unexpected caching behavior.

Specifically, we observed that when a feature flag is updated, such as changes to filters or targeting rules, the evaluation continues to return outdated results for several minutes. This suggests that the evaluation context is being cached based on the target key, resulting in stale data being served.

As a potential workaround, we considered randomly generating a unique string for the target key per request to bypass caching entirely. While this approach appears to invalidate the cache effectively, we are unsure whether it aligns with best practices for feature flagging systems.

Any advice on this?

**Comment by thomaspoignant** [comment]
>we are currently encountering an issue related to feature flag evaluation, where we do not have a dedicated target key available. In our setup, the target key property is shared across multiple users, which appears to be causing unexpected caching behavior.

@Volodymyr-Maiboroda-Imprivata which provider are you using with GO Feature Flag? just to be sure if it is or not a caching issue in one of them.

>As a potential workaround, we considered randomly generating a unique string for the target key per request to bypass caching entirely. While this approach appears to invalidate the cache effectively, we are unsure whether it aligns with best practices for feature flagging systems.

This is a totally valid way of doing it too and will not cause any problems, but it also means that the same person doing 2 different evaluations may end up with different behavior, depending on your usecase this could be valid or not.

**Comment by Volodymyr-Maiboroda-Imprivata** [comment]
@thomaspoignant we use Github
